File: exercism/python/archive/crypto-square/crypto_square.py
""" module to return the old-skool crypto square method for a given input """
# -*- coding:utf-8 -*-

from __future__ import unicode_literals

import math

def encode(src):
    """ Function to encode the given src string """
    if not src:
        return ''

    src = ''.join([x.lower() for x in src if x.isalpha() or x.isdigit()])

    rows = int(round(math.sqrt(len(src))))
    cols = int(math.ceil(len(src) / rows))

    if len(src) % rows != 0:
        cols += 1

    words = [src[i:i+cols] for i in range(0, len(src), cols)]

    return ' '.join([''.join(x[y] for x in words if y < len(x)) for y in range(cols)])

File: exercism/go/archive/crypto-square/crypto_square.go
// Package cryptosquare returns the square-coded version of a string
package cryptosquare

import (
	"bytes"
	"math"
	"strings"
	"unicode"
)

// testVersion is the test suite we're up against
const testVersion = 2

func round(f float64) float64 {
	return math.Floor(f + .5)
}

// Encode receives a decoded string and returns the encoded version using square-coding
func Encode(input string) (output string) {
	if len(input) <= 0 {
		return ""
	}

	var buf bytes.Buffer

	for _, each := range strings.ToLower(input) {
		if !unicode.IsLetter(each) && !unicode.IsNumber(each) {
			continue
		}
		buf.WriteRune(each)
	}

	input = buf.String()
	buf.Reset()

	rows := int(round(math.Sqrt(float64(len(input)))))
	cols := int(math.Ceil(float64(len(input) / rows)))

	if (len(input) % rows) != 0 {
		cols++
	}

	slicer := []string{}

	for x := 0; x < cols; x++ {
		for y := 0; y < rows && (((cols * y) + x) < (len(input))); y++ {
			buf.WriteByte(input[(cols*y)+x])
		}
		slicer = append(slicer, buf.String())
		buf.Reset()
	}

	return strings.Join(slicer, " ")
}
